pkg/operation/botanist/component/deploy.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import "context"
+
+// OpDestroy creates a DeployWaiter which calls Destroy instead of create
+// and WaitCleanup instead of Wait
+func OpDestroy(dw ...DeployWaiter) DeployWaiter {
+	return &deploy{
+		dw:     dw,
+		invert: true,
+		wait:   false,
+	}
+}
+
+// OpDestroyAndWait creates a Deployer which calls Destroy instead of create
+// and waits for destruction.
+func OpDestroyAndWait(dw ...DeployWaiter) Deployer {
+	return &deploy{
+		dw:     dw,
+		invert: true,
+		wait:   true,
+	}
+}
+
+// OpWaiter creates a Deployer which calls waits for each operation.
+func OpWaiter(dw ...DeployWaiter) Deployer {
+	return &deploy{
+		dw:     dw,
+		invert: false,
+		wait:   true,
+	}
+}
+
+// NoOp does nothing
+func NoOp() DeployWaiter { return &deploy{} }
+
+type deploy struct {
+	invert bool
+	wait   bool
+	dw     []DeployWaiter
+}
+
+func (d *deploy) Deploy(ctx context.Context) error {
+	if d.invert {
+		return d.Destroy(ctx)
+	}
+
+	for _, dw := range d.dw {
+		if dw == nil {
+			continue
+		}
+
+		if err := dw.Deploy(ctx); err != nil {
+			return err
+		}
+
+		if d.wait {
```

Comment:
Hm, why is `wait` a "global" setting here? What if I want to wait only for a few dedicated `Deploy` calls while I want to simply run `Deploy` without waiting for the others?

Reply:
This is mainly used to combine some steps which are always combined together - for example, deploy DNSProvider, wait for it to be ready, then deploy DNSEntry, wait for it to be ready:


```go
_ = g.Add(flow.Task{
		Name: "Deploying internal domain DNS record",
		Fn: flow.TaskFn(func(ctx context.Context) error {
				return component.OpWaiter(
						botanist.Shoot.Components.DNS.InternalProvider,
						botanist.Shoot.Components.DNS.InternalEntry,
				).Deploy(ctx)
		}).DoIf(wakeupRequired),
		Dependencies: flow.NewTaskIDs(wakeUpControlPlane),
})
```

Reply:
Sure, there are use-cases where this perfectly fits. I was more looking at the `Deploy` function from a general point of view. As it is now I would not be able to use it if I only want to wait for some `Deploy` steps and not for others. Wouldn't it make sense to at least allow the possibility to specify the wait behaviour per `DeployWaiter` interface in the list (optionally, default is "always wait")?

Reply:
> Sure, there are use-cases where this perfectly fits. I was more looking at the Deploy function from a general point of view. As it is now I would not be able to use it if I only want to wait for some Deploy steps and not for others. 

This is why I've separated the life-cycle operations into 4 steps `Deploy`, `Wait`, `Destroy`, `WaitCleanup`. With the helper functions `OpDestroy`, `OpDestroyAndWait`, `OpWaiter` you can combine different steps - `Deploy + Wait` and multiple `DeployWaiter` into various combinations without the need for modifying the underlying component implementation.

Reply:
OK, I see. Thanks 👍